docs/source/notes/modules.rst

```diff
@@ -0,0 +1,433 @@
+computation but is not learnable. For such cases, PyTorch provides the concept of "buffers", both "persistent"
+and "non-persistent". Following is an overview of the various types of state a module can have:
+
+* **Parameters**\ : learnable aspects of computation; contained within the ``state_dict``
+* **Buffers**\ : non-learnable aspects of computation
+
+  * **Persistent** buffers: contained within the ``state_dict`` (i.e. serialized when saving)
+  * **Non-persistent** buffers: not contained within the ``state_dict`` (i.e. left out of serialization)
+
+As a motivating example for the use of buffers, consider a simple module that maintains a running mean. We want
+the current value of the running mean to be considered part of the module's ``state_dict`` and serialized, but not
+learnable. This snippet shows how to use :func:`~torch.nn.Module.register_buffer` to accomplish this:
+
+.. code-block:: python
+
+   class RunningMean(nn.Module):
+     def __init__(self, num_features, momentum=0.9):
+       super().__init__()
+       self.momentum = momentum
+       self.register_buffer('mean', torch.zeros(num_features))
+     def forward(self, x):
+       self.mean = self.momentum * self.mean + (1.0 - self.momentum) * x
+       return self.mean
+
+Now, the current value of the running mean is considered part of the module's ``state_dict``
+and will be saved as part of the module when the module is serialized.
+
+.. code-block:: python
+
+   m = RunningMean(4)
+   for _ in range(10):
+     input = torch.randn(4)
+     m(input)
+
+   print(m.state_dict())
+   : OrderedDict([('mean', tensor([ 0.1041, -0.1113, -0.0647,  0.1515]))]))
+
+   # Serialized form will contain the 'mean' tensor
+   torch.save(m.state_dict(), 'mean.pt')
+
+   m_loaded = RunningMean(4)
+   m_loaded.load_state_dict(torch.load('mean.pt'))
+   assert(torch.all(m.mean == m_loaded.mean))
+
+As mentioned previously, buffers can be left out of the module's ``state_dict`` by marking them as non-persistent:
+
+.. code-block:: python
+
+   self.register_buffer('unserialized_thing', torch.randn(5), persistent=False)
+
+Both persistent and non-persistent buffers are included in model-wide device / dtype changes:
+
+.. code-block:: python
+
+   # Moves all module parameters and buffers to the specified device / dtype
+   m.to(device='cuda', dtype=torch.float64)
+
+Buffers of a module can be iterated over using :func:`~torch.nn.Module.buffers` or
+:func:`~torch.nn.Module.named_buffers`.
+
+For more information, check out:
+
+* Saving and loading: https://pytorch.org/tutorials/beginner/saving_loading_models.html
+* Serialization semantics: https://pytorch.org/docs/master/notes/serialization.html
+* What is a state dict? https://pytorch.org/tutorials/recipes/recipes/what_is_state_dict.html
+
+Module Hooks
+------------
+
+To ensure that simple modifications can be done easily, even for modules that you haven't written yourself,
```

Comment:
I don't really understand from this description why I would use hooks.

Reply:
Rewrote this section a bit, providing some context and useful examples.